scylla-macros/src/deserialize/value.rs

```diff
@@ -105,11 +112,248 @@ impl StructDesc {
+                    #macro_internal::DeserUdtTypeCheckErrorKind::TooFewFields {
+                        required_fields: vec![
+                            #(stringify!(#required_fields_names),)*
+                        ],
+                        present_fields: fields.iter().map(|(name, _typ)| name.clone()).collect(),
+                    }
+                );
+
+                // Verify that the field count is correct
+                if fields.len() < #required_field_count_lit {
+                    return ::std::result::Result::Err(too_few_fields());
+                }
+
+                let mut cql_field_iter = fields.iter();
+                #(
+                    #field_validations
+                )*
+
+                // All is good!
+                ::std::result::Result::Ok(())
+            }
+        }
+    }
+}
+
+struct DeserializeAssumeOrderGenerator<'sd>(&'sd StructDesc);
+
+impl<'sd> DeserializeAssumeOrderGenerator<'sd> {
+    fn generate_finalize_field(&self, field: &Field) -> syn::Expr {
+        if field.skip {
+            // Skipped fields are initialized with Default::default()
+            return parse_quote! {
+                ::std::default::Default::default()
+            };
+        }
+
+        let macro_internal = self.0.struct_attrs().macro_internal_path();
+        let cql_name_literal = field.cql_name_literal();
+        let deserializer = field.deserialize_target();
+        let constraint_lifetime = self.0.constraint_lifetime();
+
+        let deserialize: syn::Expr = parse_quote! {
+            <#deserializer as #macro_internal::DeserializeValue<#constraint_lifetime>>::deserialize(cql_field_typ, value)
+                .map_err(|err| #macro_internal::mk_value_deser_err::<Self>(
+                    typ,
+                    #macro_internal::UdtDeserializationErrorKind::FieldDeserializationFailed {
+                        field_name: #cql_name_literal.to_owned(),
+                        err,
+                    }
+                ))?
+        };
+
+        // Action performed in case of field name mismatch.
+        let name_mismatch: syn::Expr = parse_quote! {
+            panic!(
+                "type check should have prevented this scenario - field name mismatch! Rust field name {}, CQL field name {}",
+                #cql_name_literal,
+                cql_field_name
+            )
+        };
+
+        let name_check_and_deserialize: syn::Expr = parse_quote! {
+            if #cql_name_literal == cql_field_name {
+                #deserialize
+            } else {
+                #name_mismatch
+            }
+        };
+
+        parse_quote! {
+            {
+                let next_cql_field = cql_field_iter.next()
+                    .map(|(specs, value_res)| value_res.map(|value| (specs, value)))
+                    // Type check has ensured that there are enough CQL UDT fields.
+                    .expect("Too few CQL UDT fields - type check should have prevented this scenario!")
```

Comment:
OTOH, type check is done only once per query result, whereas `deserialize` is called for each row. So if `deserialize` is at least a bit faster because of not having at least part of type checking logic inside of it, the split is a win.